plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java

```diff
@@ -310,6 +316,23 @@ protected String getTableSchemaName(ResultSet resultSet)
 
             case Types.DATE:
                 return Optional.of(dateColumnMapping());
+
+            case Types.TIMESTAMP:
+                // TODO support higher precisions (https://github.com/trinodb/trino/issues/6910)
+                TimestampType timestampType = TIMESTAMP_MILLIS;
+                return Optional.of(ColumnMapping.longMapping(
+                        timestampType,
+                        (resultSet, columnIndex) -> {
+                            LocalDateTime timestamp = resultSet.getTimestamp(columnIndex).toLocalDateTime();
+                            int roundedNanos = toIntExact(round(timestamp.getNano(), 9 - timestampType.getPrecision()));
+                            LocalDateTime roundedTimestamp = timestamp
+                                    .withNano(roundedNanos % toIntExact(NANOSECONDS_PER_SECOND))
+                                    .plusSeconds(roundedNanos / NANOSECONDS_PER_SECOND);
+                            return toPrestoTimestamp(timestampType, roundedTimestamp);
```

Comment:
The need for rounding applies to MemSQL, Druid, Redshift connectors as well -- i.e. caller of https://github.com/trinodb/trino/blob/922bfa1a32cc2e39d24d79f9801099938614dd82/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java#L634 who do not handle timestamps on their own.


That flow used to round until be2b97a8480f657f311e673fa38867c8faf4e2df, but it went unnoticed until we added more verifications in other places.

MySQL is not a special case here, and we have other connectors suffering -- those who're not smart enough to handle their mapping by themsleves. We should fix the mapping returned from `io.trino.plugin.jdbc.StandardColumnMappings#jdbcTypeToPrestoType` by rounding to millis. This should be an opt-in, ie it must not be applied to `timestampColumnMappingUsingSqlTimestamp` or `timestampColumnMapping` when someone calls them directly.

Reply:
continued @ https://github.com/trinodb/trino/pull/6893#discussion_r576641252

Reply:
Makes sense. I'll change the mapping returned from `legacyToPrestoType` while leaving the type-level mappings (e.g. `timestampColumnMapping` as it is.

Reply:
With rounding enabled all `verifyPredicate` calls fail for values where we round during reads - most likely because to predicate generated in `verifyPredicate` is being pushed down.

I'll need to disable pushdown for all connectors using the `legacyToPrestoType`.